Report line number of test when should_panic test failed

xizheyin opened the pull request:
Closes #137405 

---

try-job: x86_64-gnu-llvm-19-3
try-job: test-various

xizheyin commented:
r? @jieyouxu

xizheyin committed:
Report span of test when should_panic test failed

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,30 @@
+#[test]
+#[should_panic = "SubString"]
+fn do_panic2() {
+    // DON'T PANIC
```
```suggestion
```

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[test]
+#[should_panic]
+fn do_panic1() {
```
```suggestion
fn should_panic_with_any_message() {
```

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[test]
+#[should_panic = "SubString"]
+fn do_panic2() {
```
```suggestion
fn should_panic_with_substring() {
```

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[test]
+#[should_panic]
+fn dont_panic1() {
```
```suggestion
fn should_panic_with_any_message_does_not_panic() {
```

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,30 @@
+
+#[test]
+#[should_panic = "SubString"]
+fn dont_panic2() {
```
```suggestion
fn should_panic_with_substring_does_not_panic() {
```

joshka commented on tests/ui/test-should-panic-failed-show-span.rs:
Do you need to add a test for panicking with the wrong string?

```rust
#[test]
#[should_panic = "Substring"]
fn should_panic_with_substring_panics_with_incorrect_string() {
    panic!("ZOMGWTFBBQ");
}
```

joshka commented on library/test/src/test_result.rs:
Do the parts on line 58 to 68 also need to have the span added (and be tested). Not covered in the linked issue, but seems like these would be also good to fix in this PR if they're currently incorrect.

joshka reviewed (commented):
Test names are nits, but make the reason for each test self describing.

I think it probably makes sense to also fix the substring tests if they're also broken (I haven't manually checked those messages to see if there's span info coming from elsewhere in the stack, but these seem like they should be tested with the same tests as are introduced here).

xizheyin commented on tests/ui/test-should-panic-failed-show-span.run.stdout:
```diff
@@ -0,0 +1,40 @@
+
+thread 'should_panic_with_substring_panics_with_incorrect_string' panicked at $DIR/test-should-panic-failed-show-span.rs:34:5:
+ZOMGWTFBBQ
+note: should panic with message at $DIR/test-should-panic-failed-show-span.rs:33:4
```
Now, I add support for cases that should panic with message but panic with incorrect string or non-string value.

joshka commented on library/test/src/test_result.rs:
```diff
@@ -45,6 +45,18 @@ pub(crate) fn calc_result<'a>(
     time_opts: Option<&time::TestTimeOptions>,
     exec_time: Option<&time::TestExecTime>,
 ) -> TestResult {
+    let span_str = if desc.source_file != "" {
```
Nit: what about using a name like?
```suggestion
    let at_location = if desc.source_file != "" {
```

joshka commented on library/test/src/test_result.rs:
```diff
@@ -45,6 +45,18 @@ pub(crate) fn calc_result<'a>(
+        "".to_string()
+    };
+
+    let span_should_panic_with_message_didnot_panic = if desc.source_file != "" {
```
Nit: Given this is a local variable, I'd generally choose a bit of a shorter name. It's also only applicable to one of the match arms, so move it within that block below (at line 63)

joshka reviewed (approved):
Generally LGTM - small nits

(Note: I'm not a member of the rust project, so this likely still requires approval from someone who is)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13913252839/job/38931600820) [(plain)](https://github.com/rust-lang/rust/commit/c7b4523858b1ab0df169a825c9b54ec8d84d7159/checks/38931600820/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#19 exporting to docker image format
#19 sending tarball 20.2s done
#19 DONE 32.9s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
  Number of decisions:   4447
  longest path:          1159 (code:    152)
  longest backtrack:       66 (code:    428)
Shared 86733 out of 152951 states by creating 14756 new states, saving 71977
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc: In function ‘rtx_def* extract_bit_field_1(rtx, poly_uint64, poly_uint64, int, rtx, machine_mode, machine_mode, bool, bool, rtx_def**)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1864:45: warning: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ may be used uninitialized [-Wmaybe-uninitialized]
 1864 |       rtx sub = extract_bit_field_as_subreg (mode1, op0, imode,
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
 1865 |                                              bitsize, bitnum);
      |                                              ~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/expmed.cc:1824:19: note: ‘*(unsigned int*)((char*)&imode + offsetof(scalar_int_mode, scalar_int_mode::m_mode))’ was declared here
 1824 |   scalar_int_mode imode;
      |                   ^~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc: In member function ‘void range_def_chain::dump(FILE*, basic_block, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gimple-range-gori.cc:319:19: warning: format not a string literal and no format arguments [-Wformat-security]
  319 |           fprintf (f, prefix);
      |           ~~~~~~~~^~~~~~~~~~~
---
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:7930:9: warning: ignoring return value of ‘ssize_t write(int, const void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 7930 |   write (fd, "\n\n", 2);
      |   ~~~~~~^~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc: In member function ‘void driver::final_actions() const’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/gcc.cc:9307:13: warning: ignoring return value of ‘int truncate(const char*, __off_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 9307 |     truncate(totruncate_file, 0);
      |     ~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc: In function ‘bool find_and_merge_options(int, off_t, const char*, vec<cl_decoded_option>, bool, vec<cl_decoded_option>*, const char*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto-wrapper.cc:1165:8: warning: ignoring return value of ‘ssize_t read(int, void*, size_t)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 1165 |   read (fd, data, length);
      |   ~~~~~^~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc: In function ‘void lto_resolution_read(splay_tree, FILE*, lto_file*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2091:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2091 |   fscanf (resolution, " ");   /* Read white space.  */
      |   ~~~~~~~^~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2093:9: warning: ignoring return value of ‘size_t fread(void*, size_t, size_t, FILE*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2093 |   fread (obj_name, sizeof (char), name_len, resolution);
      |   ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/lto/lto-common.cc:2113:10: warning: ignoring return value of ‘int fscanf(FILE*, const char*, ...)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
 2113 |   fscanf (resolution, "%u", &num_symbols);
      |   ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/gcc/jit/jit-recording.cc:32:
---
Applying io_quotes_use            to linux/blkzoned.h
Applying io_quotes_use            to linux/ipmi.h
Applying io_quotes_use            to linux/psp-dbc.h
Applying io_quotes_use            to linux/bt-bmc.h
Applying io_quotes_use            to linux/tps6594_pfsm.h
Applying io_quotes_use            to linux/cxl_mem.h
Applying io_quotes_use            to linux/wmi.h
Applying io_quotes_use            to linux/auto_fs.h
Applying io_quotes_use            to linux/mmtimer.h
Applying io_quotes_use            to linux/f2fs.h
Applying io_quotes_use            to linux/vhost.h
---
Applying machine_name             to x86_64-linux-gnu/bits/unistd_ext.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/mtrr.h
Applying io_quotes_use            to x86_64-linux-gnu/asm/amd_hsmp.h
Applying machine_name             to openssl/e_os2.h
Applying io_quotes_use            to drm/xe_drm.h
Applying io_quotes_use            to drm/radeon_drm.h
Applying io_quotes_use            to drm/panfrost_drm.h
Applying io_quotes_use            to drm/etnaviv_drm.h
Applying io_quotes_use            to drm/lima_drm.h
Applying io_quotes_use            to drm/qaic_accel.h
Applying io_quotes_use            to drm/vc4_drm.h
Applying io_quotes_use            to drm/i915_drm.h
Applying io_quotes_use            to drm/omap_drm.h
Applying io_quotes_use            to drm/pvr_drm.h
Applying io_quotes_use            to drm/amdgpu_drm.h
Applying io_quotes_use            to drm/vgem_drm.h
Applying io_quotes_use            to drm/msm_drm.h
Applying io_quotes_use            to drm/v3d_drm.h
Applying io_quotes_use            to drm/exynos_drm.h
Applying io_quotes_use            to drm/nouveau_drm.h
Applying io_quotes_use            to drm/drm.h
Applying io_quotes_use            to drm/habanalabs_accel.h
Applying io_quotes_use            to drm/tegra_drm.h
Applying io_quotes_use            to rdma/rdma_user_ioctl.h
cc1: note: self-tests are not enabled in this build
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc: In function ‘void server(bool, int, module_resolver*)’:
/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/src/c++tools/server.cc:620:10: warning: ignoring return value of ‘int pipe(int*)’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
---
[RUSTC-TIMING] proc_macro test:false 5.902
[RUSTC-TIMING] test test:false 9.248
    Finished `release` profile [optimized] target(s) in 1m 20s
##[endgroup]
[TIMING] core::build_steps::compile::Std { target: x86_64-unknown-linux-gnu, compiler: Compiler { stage: 1, host: x86_64-unknown-linux-gnu, forced_compiler: true }, crates: [], force_recompile: false, extra_rust_args: ["-Csymbol-mangling-version=v0", "-Cpanic=abort"], is_for_mir_opt_tests: false } -- 80.612
Testing GCC stage1 (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
 Downloading crates ...
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
[RUSTC-TIMING] boml test:false 0.819
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
[RUSTC-TIMING] y test:false 2.981
    Finished `release` profile [optimized] target(s) in 4.23s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-backend gcc --gcc-path /checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
`--gcc-path` was provided, ignoring config file. Using `/checkout/obj/build/x86_64-unknown-linux-gnu/gcc/install/lib` as path for libgccjit
[BUILD] mini_core
[RUSTC-TIMING] mini_core test:false 0.198
[BUILD] example
[AOT] mini_core_hello_world
[RUSTC-TIMING] mini_core_hello_world test:false 0.186
---
diff of stdout:

5 failures:
6 
7 ---- $DIR/failed-doctest-should-panic.rs - Foo (line 10) stdout ----
- note: test did not panic as expected
+ note: test did not panic as expected at $DIR/failed-doctest-should-panic.rs:10:0
9 
10 failures:
11     $DIR/failed-doctest-should-panic.rs - Foo (line 10)
---
To only update this specific test, also pass `--test-args doctest/failed-doctest-should-panic.rs`

error: 1 errors occurred comparing output.
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "/checkout/tests/rustdoc-ui/doctest/failed-doctest-should-panic.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(test,FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-ui/doctest/failed-doctest-should-panic" "-A" "internal_features" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2024" "--test"
--- stdout -------------------------------

running 1 test
test /checkout/tests/rustdoc-ui/doctest/failed-doctest-should-panic.rs - Foo (line 10) - should panic ... FAILED

```

</details>

jieyouxu commented:
> (Note: I'm not a member of the rust project, so this likely still requires approval from someone who is)

Though reviews are still very much appreciated 😁

bors commented:
:hourglass: Trying commit a164290333f50fc7cf373f89b37894f4a8ad6613 with merge b7cffdfcf2ab6d783e833e46b5dd168197dbcd3e...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a164290333f50fc7cf373f89b37894f4a8ad6613","merge_sha":"b7cffdfcf2ab6d783e833e46b5dd168197dbcd3e"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13933401439/job/39001638743)
Build commit: b7cffdfcf2ab6d783e833e46b5dd168197dbcd3e (`b7cffdfcf2ab6d783e833e46b5dd168197dbcd3e`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13933401439/job/39001638743"},"merge_sha":"b7cffdfcf2ab6d783e833e46b5dd168197dbcd3e"} -->

joboet commented:
I've never touched the `test` crate before so I don't feel qualified to review this.
r? @thomcc 
as you are a member of testing-devex

thomcc commented on library/test/src/test_result.rs:
```diff
@@ -53,26 +59,32 @@ pub(crate) fn calc_result<'a>(
                 .map(|e| &**e)
                 .or_else(|| err.downcast_ref::<&'static str>().copied());
 
+            let shoud_panic_location = if desc.source_file != "" {
```
Typo nit: `should_panic_location`.

xizheyin replied:
Thanks

Kobzol commented:
> @bors try
> 
> Before the change fails snapshots in Cargo's test testsuite

(Just noting that this doesn't run any Cargo tests, in fact it doesn't run any tests at all :) ).

epage commented on library/test/src/test_result.rs:
```diff
@@ -53,26 +59,32 @@ pub(crate) fn calc_result<'a>(
                 .map(|e| &**e)
                 .or_else(|| err.downcast_ref::<&'static str>().copied());
 
+            let should_panic_location = if desc.source_file != "" {
```
Should we instead check for `!at_location.is_empty()`?

epage commented on library/test/src/test_result.rs:
```diff
@@ -53,26 +59,32 @@ pub(crate) fn calc_result<'a>(
-                    r#"panic did not contain expected string
+                    r#"{should_panic_location}panic did not contain expected string
       panic message: `{panic_str:?}`,
- expected substring: `{msg:?}`"#
+      expected substring: `{msg:?}`"#
```
This is an unrelated change and should be split out to a different commit, maybe even a different PR.

Also, unsure if we should do it.

epage commented on library/test/src/test_result.rs:
```diff
@@ -53,26 +59,32 @@ pub(crate) fn calc_result<'a>(
- found non-string value: `{:?}`
-     expected substring: `{:?}`"#,
+                    r#"{should_panic_location}expected panic with string value,
+      found non-string value: `{:?}`
+      expected substring: `{:?}`"#,
```
ditto about alignment changes

epage commented on library/test/src/test_result.rs:
```diff
@@ -45,6 +45,12 @@ pub(crate) fn calc_result<'a>(
     time_opts: Option<&time::TestTimeOptions>,
     exec_time: Option<&time::TestExecTime>,
 ) -> TestResult {
+    let at_location = if desc.source_file != "" {
```
For myself, I'd be a bit more specific, like `fn_location` or `test_location`

chenyukang commented:
All comment feedbacks were resolved, LGTM, thanks!
@bors r=chenyukang

bors commented:
:pushpin: Commit cc2ed5ae136fc1b0f6a0132b9623d5a837ded306 has been approved by `chenyukang`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=chenyukang cc2ed5ae136fc1b0f6a0132b9623d5a837ded306 -->
<!-- homu: {"type":"Approved","sha":"cc2ed5ae136fc1b0f6a0132b9623d5a837ded306","approver":"chenyukang","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

xizheyin commented:
Why did it cause a failure? I noticed that the failed stdout seems to have extra backtrace inside it.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-19-3`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14308845672/job/40098889959) [(plain)](https://github.com/rust-lang-ci/rust/commit/909f4fc4fb2fb6e5ef0c72acf71d798b6bc5f4ad/checks/40098889959/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Saved the actual run.stdout to "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/test-should-panic-failed-show-span/test-should-panic-failed-show-span.run.stdout"
diff of run.stdout:

17 
18 thread 'should_panic_with_substring_panics_with_incorrect_string' panicked at $DIR/test-should-panic-failed-show-span.rs:34:5:
19 ZOMGWTFBBQ
+ stack backtrace:
+    0: std::panicking::begin_panic
+    1: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_incorrect_string
+    2: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_incorrect_string::{{closure}}
+    3: core::ops::function::FnOnce::call_once
+ note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
20 note: panic did not contain expected string
21       panic message: `"ZOMGWTFBBQ"`,
22  expected substring: `"message"`

24 
25 thread 'should_panic_with_substring_panics_with_non_string_value' panicked at $DIR/test-should-panic-failed-show-span.rs:40:5:
26 Box<dyn Any>
+ stack backtrace:
+    0: std::panicking::begin_panic
+    1: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_non_string_value
+    2: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_non_string_value::{{closure}}
+    3: core::ops::function::FnOnce::call_once
+ note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
27 note: expected panic with string value,
28  found non-string value: `TypeId(0x56ced5e4a15bd89050bb9674fa2df013)`
29      expected substring: `"message"`

34     should_panic_with_substring_panics_with_incorrect_string
35     should_panic_with_substring_panics_with_non_string_value
36 
- test result: FAILED. 2 passed; 4 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
---
The actual run.stdout differed from the expected run.stdout.

error: 1 errors occurred comparing run output.
status: exit status: 101
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/test-should-panic-failed-show-span" && RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" RUST_TEST_THREADS="4" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/test-should-panic-failed-show-span/a" "--test-threads=1"
--- stdout -------------------------------

running 6 tests
test should_panic_with_any_message - should panic ... ok
test should_panic_with_any_message_does_not_panic - should panic ... FAILED
---
   1: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_incorrect_string
   2: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_incorrect_string::{{closure}}
   3: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
note: panic did not contain expected string
      panic message: `"ZOMGWTFBBQ"`,
 expected substring: `"message"`
---- should_panic_with_substring_panics_with_non_string_value stdout ----

thread 'should_panic_with_substring_panics_with_non_string_value' panicked at /checkout/tests/ui/test-should-panic-failed-show-span.rs:40:5:
Box<dyn Any>
stack backtrace:
   0: std::panicking::begin_panic
   1: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_non_string_value
   2: test_should_panic_failed_show_span::should_panic_with_substring_panics_with_non_string_value::{{closure}}
   3: core::ops::function::FnOnce::call_once
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
note: expected panic with string value,
 found non-string value: `TypeId(0x56ced5e4a15bd89050bb9674fa2df013)`
     expected substring: `"message"`

failures:
    should_panic_with_any_message_does_not_panic
    should_panic_with_message_does_not_panic
    should_panic_with_substring_panics_with_incorrect_string
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14308845672/job/40101733535)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14308845672/job/40101733535","builder_name":"checks-actions"} -->

chenyukang commented:
> Why did it cause a failure? I noticed that the failed stdout seems to have extra backtrace inside it.

maybe caused by some other changes from upstream, can you rebase this PR and have a testing on local, which may can reproduce the error.

Zalathar commented:
It looks like other run-fail tests for test attributes use `//@ exec-env:RUST_BACKTRACE=0` to explicitly disable backtraces.

Zalathar commented on tests/ui/test-should-panic-failed-show-span.rs:
Suggestion: There are other tests for the `#[test]` and `#[should_fail]` attributes in the `tests/ui/test-attrs/` directory, so I think this test should be in that subdirectory instead of the very crowded `tests/ui/` directory.

xizheyin commented:
What's werid is that when I updated the latest code, it didn't fail in the local test, so I suspect that the PR that happened to rollup together had content that conflicted with this.

However that's ok, I've added `//@ exec-env: RUST_BACKTRACE=0` to avoid backtrace.

Thank both of you, @chenyukang @Zalathar

xizheyin commented:
It seems that there are too many (6) functions in the test, causing the execution time to become 0.01s sometimes, I deleted the tests that have nothing to do with the change, and only kept the two tests that “should be panic but are not”.

Zalathar commented:
You will also need this to prevent the test time from causing unpredictable failures:

```rust
//@ normalize-stdout: "finished in \d+\.\d+s" -> "finished in $$TIME"
```

bors commented:
:hourglass: Trying commit 92c0fc45e8bf10097e4bf6a9b5f4818f7a142cfb with merge e3d12c70105892942a1e11612422c4dd5870277f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"92c0fc45e8bf10097e4bf6a9b5f4818f7a142cfb","merge_sha":"e3d12c70105892942a1e11612422c4dd5870277f"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14312112185/job/40112321578)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14312112185/job/40112321578","builder_name":"checks-actions"} -->

xizheyin commented:
Oops, still failed.

xizheyin commented:
```
//@ run-fail
//@ check-run-results
//@ compile-flags: --test
//@ exec-env:RUST_BACKTRACE=0
//@ normalize-stdout: "finished in \d+\.\d+s" -> "finished in $$TIME"
//@ run-flags: --test-threads=1
```
Now maybe it's okay. I learned a lot and next time I can look for the appropriate directives in similar tests.

Zalathar commented:
@bors try

bors commented:
:hourglass: Trying commit 5a1322580f69398f881c36759dcd11992345a6ce with merge 183e9d447d46102eff75d541478457051524346d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"5a1322580f69398f881c36759dcd11992345a6ce","merge_sha":"183e9d447d46102eff75d541478457051524346d"} -->

rust-log-analyzer commented:
The job **`test-various`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/14321282157/job/40138596025) [(plain)](https://github.com/rust-lang-ci/rust/commit/183e9d447d46102eff75d541478457051524346d/checks/40138596025/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.url=https://github.com/rust-lang/rust
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
- ---- should_panic_with_message_does_not_panic stdout ----
- note: test did not panic as expected at $DIR/test-should-panic-failed-show-span.rs:28:4
- ---- should_panic_with_substring_panics_with_incorrect_string stdout ----
17 
- thread 'should_panic_with_substring_panics_with_incorrect_string' panicked at $DIR/test-should-panic-failed-show-span.rs:35:5:
- ZOMGWTFBBQ
- note: panic did not contain expected string
-       panic message: `"ZOMGWTFBBQ"`,
-  expected substring: `"message"`
- ---- should_panic_with_substring_panics_with_non_string_value stdout ----
24 
- thread 'should_panic_with_substring_panics_with_non_string_value' panicked at $DIR/test-should-panic-failed-show-span.rs:41:5:
- Box<dyn Any>
- note: expected panic with string value,
-  found non-string value: `TypeId(0x56ced5e4a15bd89050bb9674fa2df013)`
-      expected substring: `"message"`
30 
- failures:
-     should_panic_with_any_message_does_not_panic
-     should_panic_with_message_does_not_panic
-     should_panic_with_substring_panics_with_incorrect_string
---
The actual run.stdout differed from the expected run.stdout.

error: 1 errors occurred comparing run output.
status: exit status: 0
command: cd "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/test-attrs/test-should-panic-failed-show-span" && RUSTC="/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" RUST_BACKTRACE="0" RUST_TEST_THREADS="4" "/wasmtime-v19.0.0-x86_64-linux/wasmtime" "run" "-C" "cache=n" "--dir" "." "--env" "RUSTC_BOOTSTRAP" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/test-attrs/test-should-panic-failed-show-span/a.wasm" "--test-threads=1"
--- stdout -------------------------------

running 6 tests
test should_panic_with_any_message - should panic ... ignored
test should_panic_with_any_message_does_not_panic - should panic ... ignored
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14321282157/job/40140466767)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/14321282157/job/40140466767","builder_name":"checks-actions"} -->

xizheyin committed:
Adjust test directives

Signed-off-by: xizheyin <xizheyin@smail.nju.edu.cn>

chenyukang commented:
@bors try

bors commented:
:hourglass: Trying commit 4f5edfd7a1d39f84de36ca8747607c531ea22b43 with merge 1fa8aae57717ef073c2594d06714f5e1bf793ffc...
<!-- homu: {"type":"TryBuildStarted","head_sha":"4f5edfd7a1d39f84de36ca8747607c531ea22b43","merge_sha":"1fa8aae57717ef073c2594d06714f5e1bf793ffc"} -->

jieyouxu commented on tests/ui/test-attrs/test-should-panic-failed-show-span.rs:
```diff
@@ -1,9 +1,11 @@
+//@ no-prefer-dynamic
```
Suggestion: I think this isn't needed

xizheyin replied:
I removed it, I was more curious at what point it would be useful.

jieyouxu commented on tests/ui/test-attrs/test-should-panic-failed-show-span.rs:
```diff
@@ -1,9 +1,11 @@
 //@ exec-env:RUST_BACKTRACE=0
 //@ normalize-stdout: "finished in \d+\.\d+s" -> "finished in $$TIME"
-//@ run-flags: --test-threads=1
+//@ needs-threads
```
Suggestion: I *think* for libtest unwinding you might need sth like

```
//@ needs-unwind (panic)
```

Zalathar replied:
I think `--test-threads=1` might still be needed for deterministic ordering of test results.

jieyouxu replied:
Oh I didn't notice that being removed. Yes, `--test-threads=1` has to be used for determinism.

xizheyin replied:
Now, I added it.

Zalathar commented on tests/ui/test-attrs/test-should-panic-failed-show-span.rs:
```diff
@@ -0,0 +1,44 @@
+    panic!("ZOMGWTFBBQ");
+}
+
+#[test]
+#[should_panic = "message"]
+fn should_panic_with_substring_panics_with_non_string_value() {
+    panic!(123); //~ WARNING panic message is not a string literal
+}
```
We can add `#[expect(non_fmt_panics)]` to this function to get rid of the compiler warning and simplify the output snapshots, since the warning isn't important to the test.

xizheyin replied:
Yeah, I added it.

xizheyin commented on tests/ui/test-attrs/test-should-panic-failed-show-span.run.stdout:
```diff
@@ -0,0 +1,32 @@
+ expected substring: `"message"`
+---- should_panic_with_substring_panics_with_non_string_value stdout ----
+note: expected panic with string value,
+ found non-string value: `TypeId(0x56ced5e4a15bd89050bb9674fa2df013)`
```
Now, it is like
```
 found non-string value: `TypeId($HEX)`
```

jieyouxu commented:
@rustbot ready

chenyukang commented:
@bors r=chenyukang

bors commented:
:pushpin: Commit ee3b557b2dff25ada8cb6248591d6f958ca2d567 has been approved by `chenyukang`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=chenyukang ee3b557b2dff25ada8cb6248591d6f958ca2d567 -->
<!-- homu: {"type":"Approved","sha":"ee3b557b2dff25ada8cb6248591d6f958ca2d567","approver":"chenyukang","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #139746) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

chenyukang commented:
@bors r=chenyukang

bors commented:
:pushpin: Commit dc3a586eed198548316fcbfff2d5c28df0728750 has been approved by `chenyukang`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=chenyukang dc3a586eed198548316fcbfff2d5c28df0728750 -->
<!-- homu: {"type":"Approved","sha":"dc3a586eed198548316fcbfff2d5c28df0728750","approver":"chenyukang","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit dc3a586eed198548316fcbfff2d5c28df0728750 with merge 07d3fd1d9b9c1f07475b96a9d168564bf528db68...
<!-- homu: {"type":"BuildStarted","head_sha":"dc3a586eed198548316fcbfff2d5c28df0728750","merge_sha":"07d3fd1d9b9c1f07475b96a9d168564bf528db68"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/14447014753/job/40523535148)
Approved by: chenyukang
Pushing 07d3fd1d9b9c1f07475b96a9d168564bf528db68 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"chenyukang","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/14447014753/job/40523535148"},"merge_sha":"07d3fd1d9b9c1f07475b96a9d168564bf528db68"} -->

rust-timer commented:
Finished benchmarking commit (07d3fd1d9b9c1f07475b96a9d168564bf528db68): [comparison URL](https://perf.rust-lang.org/compare.html?start=c580c498a1fe144d7c5b2dfc7faab1a229aa288b&end=07d3fd1d9b9c1f07475b96a9d168564bf528db68&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=c580c498a1fe144d7c5b2dfc7faab1a229aa288b&end=07d3fd1d9b9c1f07475b96a9d168564bf528db68&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=c580c498a1fe144d7c5b2dfc7faab1a229aa288b&end=07d3fd1d9b9c1f07475b96a9d168564bf528db68&stat=max-rss)
<details>
<summary>Results (primary -0.5%, secondary -2.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.5% | [-0.7%, -0.4%] | 5     |
| Improvements ✅ <br /> (secondary) | -2.2% | [-2.2%, -2.2%] | 1     |
| All ❌✅ (primary)                 | -0.5% | [-0.7%, -0.4%] | 5     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=c580c498a1fe144d7c5b2dfc7faab1a229aa288b&end=07d3fd1d9b9c1f07475b96a9d168564bf528db68&stat=cycles:u)
<details>
<summary>Results (primary -0.9%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.5%, 0.5%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.3%, -2.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.9% | [-2.3%, 0.5%]  | 2     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=c580c498a1fe144d7c5b2dfc7faab1a229aa288b&end=07d3fd1d9b9c1f07475b96a9d168564bf528db68&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 784.038s -> 784.108s (0.01%)
**Artifact size**: 365.12 MiB -> 365.13 MiB (0.00%)
